cpp/src/cider/exec/operator/join/CiderChainedHashTable.cpp

```diff
@@ -0,0 +1,136 @@
+ * under the License.
+ */
+
+#pragma once
+
+namespace cider_hashtable {
+template <typename Key,
+          typename Value,
+          typename Hash,
+          typename KeyEqual,
+          typename Grower,
+          typename Allocator>
+template <typename K, typename V>
+bool ChainedHashTable<Key, Value, Hash, KeyEqual, Grower, Allocator>::emplace_impl(
+    const K& key,
+    const V& value) {
+  size_t idx = key_to_idx(key);
+  table_key<K> inner_key = {key, true, 0};
+  buckets_[idx].push_back({inner_key, value});
+  size_++;
+  return true;
+}
+
+// TODO: assert key and value types
+template <typename Key,
+          typename Value,
+          typename Hash,
+          typename KeyEqual,
+          typename Grower,
+          typename Allocator>
+void ChainedHashTable<Key, Value, Hash, KeyEqual, Grower, Allocator>::
+    merge_other_hashtables(
+        const std::vector<std::shared_ptr<
+            BaseHashTable<Key, Value, Hash, KeyEqual, Grower, Allocator>>>& otherTables) {
+  for (const auto& table_ptr_tmp : otherTables) {
+    ChainedHashTable* table_ptr = dynamic_cast<ChainedHashTable*>(table_ptr_tmp.get());
+    auto other_table_bucket = table_ptr->get_buckets();
+    size_ += table_ptr->size();
+    size_type bucket_size = bucket_count();
+    for (int i = 0; i < bucket_size; i++) {
+      buckets_[i].insert(
+          buckets_[i].end(), other_table_bucket[i].begin(), other_table_bucket[i].end());
+    }
+  }
+}
+
+template <typename Key,
+          typename Value,
+          typename Hash,
+          typename KeyEqual,
+          typename Grower,
+          typename Allocator>
+template <typename K>
+bool ChainedHashTable<Key, Value, Hash, KeyEqual, Grower, Allocator>::contains_impl(
+    const K& key) {
+  size_t idx = key_to_idx(key);
+  std::vector<value_type> slot = buckets_[idx];
+  for (auto element : slot) {
+    if (element.first.key == key) {
+      return true;
+    }
+  }
+  return false;
+}
+
+// todo: set an empty value
+template <typename Key,
+          typename Value,
+          typename Hash,
+          typename KeyEqual,
+          typename Grower,
+          typename Allocator>
+template <typename K>
+Value ChainedHashTable<Key, Value, Hash, KeyEqual, Grower, Allocator>::find_impl(
+    const K& key) {
+  size_t idx = key_to_idx(key);
+  std::vector<value_type> slot = buckets_[idx];
+  for (auto element : slot) {
+    if (element.first.key == key) {
+      return element.second;
+    }
+  }
+  return Value();
+}
+
+template <typename Key,
+          typename Value,
+          typename Hash,
+          typename KeyEqual,
+          typename Grower,
+          typename Allocator>
+template <typename K>
+std::vector<Value>
+ChainedHashTable<Key, Value, Hash, KeyEqual, Grower, Allocator>::find_all_impl(
+    const K& key) {
+  std::vector<Value> result;
+  size_t idx = key_to_idx(key);
+  std::vector<value_type> slot = buckets_[idx];
+  for (auto element : slot) {
```

Comment:
Q: When some values belong to a same key, will it be better if we put all values to a single key slot instead of storing multiple pairs? I mean there will only be one mapping relation of a key in HashTable.

Reply:
This is a trade off between write and read. Making them together is good for probing but needs extra step when build.